fv3core/stencils/fvtp2d.py

```diff
@@ -75,6 +76,62 @@ def transport_flux_xy(
             fy = transport_flux(fy, fy2, mfy)
 
 
+@dataclasses.dataclass
+class CopiedCorners:
+    """
+    Data container for storages with corners copied for differencing
+    along the x- and y-directions.
+
+    Attributes:
+        base: writeable version of storage with no guarantees about corner data
+        x_differenceable: read-only version of storage which can be differenced
+            along the x-direction
+        y_differenceable: read-only version of storage which can be differenced
+            along the y-direction
+    """
+
+    base: FloatField
+    x_differenceable: FloatField
+    y_differenceable: FloatField
+
+
+class PreAllocatedCopiedCornersFactory:
+    """
+    Creates CopiedCorners from a field, using an init-compiled stencil
+    and pre-allocated output fields.
+    """
+
+    def __init__(
+        self,
+        grid_indexing: GridIndexing,
+        *,
+        dims: Sequence[str],
+        y_temporary: FloatField,
+    ):
+        """
+        Args:
+            grid_indexing: information about grid size
+            dims: dimensionality of data to be copied
+            y_temporary: if given, storage to use for y-differenceable field
+                (x-differenceable field uses same memory as base storage),
+                if None then a storage is initialized based on max shape
+        """
+        if y_temporary is None:
+            y_temporary = utils.make_storage_from_shape(
```

Comment:
Could the `y_temporary` also use the same storage? Probably triple-aliasing is not a good idea, just curious.